h2o-core/src/main/java/hex/grid/filter/PermutationFilter.java

```diff
@@ -0,0 +1,25 @@
+package hex.grid.filter;
+
+import hex.Model;
+
+/**
+ * Provides interface required to enable filtering within {@link hex.grid.HyperSpaceWalker}
+ * 
+ * @param <MP> as {@link hex.grid.HyperSpaceWalker} is working only with descendants of the {@link Model.Parameters},
+ *            this interface is generic with corresponding type parameter
+ */
+public interface PermutationFilter<MP extends Model.Parameters> {
```

Comment:
As mentioned above, you should be able to reuse the same interface as for what you call `FilterFunction`, although if you need to keep one, I vote for `__Filter` against `__FilterFunction`.

Reply:
@seb-h2o trying to make things that are not completely the same to look similar(extend same interface) can imp make things even more confusing. For example here we don't need `activate` method for our PermutationFilter.

Reply:
@seb-h2o and the main thing is that we need two methods: one (let's say `test`) that checks whether provided filter function(condition clause) matches permutation, another one that will return higher level test result ( based on state of the function). So that is why I need two levels of abstraction.

Reply:
> For example here we don't need activate method for our PermutationFilter

correct, and I think I see your point, but it makes `PermutationFilter` even simpler, but still connected with the other filter functions, somehow:
```
@FunctionalInterface
interface PermutationFilter
    boolean match(MP)
    default void reset() {}

interface TwoStepsPermutationFilter extends PermutationFilter
     default void accept(boolean globalMatch, MP params) {}
```
the second one is just a special case: when you know you may need stateful ones, you require `TwoStepsPermutationFilter`, otherwise you just require `PermutationFilter`.

> main thing is that we need two methods

note that with the `accept` proposal, when needed, the internal `test` function can be private.

Reply:
@seb-h2o wait a bit. I will commit updated version. It will move conversation further.

Reply:
@seb-h2o what I don't like in this `boolean globalMatch` parameter is that function should care about outside world. Even though it is always the case for function's inputs but here it requires extra context for understanding which I think is bad.  It is even difficult to write javadoc for this parameter. We at once need to provide links to other classes where this input parameter will come from. In other words bottom-up approach will not relieve users from necessity to know more about other entities.

Reply:
it definitely can't cover all the various filter implementations we can imagine, but I don't think we should worry about  this for now.
the main benefit currently is to make the logic behind `KeepOnlyFirstMatchFilter` completely encapsulated in its class.

Reply:
@seb-h2o we can't do with you suggestion to use same FilterFunction interface. We will not be able to distinguish two different scenarios:
1) when we pass `KeepOnlyFirstMatchFilter ` directly to walker's constructor. In this case we would want filter to decide whether it is activated itself ( only from within `test` method).
2) when we pass PermutationFilter which can combine multiple filters ( including `KeepOnlyFirstMatchFilter` ). In that case decision on activation is outside of the `KeepOnlyFirstMatchFilter` and we don't want activation logic in test method.